colossalai/inference/batch_bucket.py

```diff
@@ -0,0 +1,459 @@
+
+        Args:
+            seq (Sequence): The sequence to be added to the batch
+            alloc_block_table (torch.Tensor): The block tables to be copied and used for the sequence
+            alloc_block_table_fn (Callable[[torch.Tensor, int], None]): The function to allocate blocks for the sequence,
+                which is expected to reserve blocks and update status of kv-cache manager.
+
+        Returns:
+            block_table (torch.Tensor): The block table of the added sequence, used for block allocation in kv-cache manager.
+                None if the sequence cannot be added.
+        """
+        block_table = None
+        # TODO might consider sorting by length
+        if self._current_batch_size < self.max_batch_size:
+            self._sequences_dict[seq.request_id] = seq
+            self._sequences_indexes[seq.request_id] = self._current_batch_size
+            self._sequence_lengths[self._current_batch_size] = seq.sentence_len
+            # NOTE the added seq still require block table allocation by kvcache manager
+            block_table = self._block_tables[self._current_batch_size - 1]
+            if alloc_block_table is not None:
+                # copy block ids from provided block tables
+                self._block_tables[self._current_batch_size - 1] = alloc_block_table
+            elif alloc_block_table_fn:
+                alloc_block_table_fn(block_table, self._sequence_lengths[self._current_batch_size - 1].item())
+            self._current_batch_size += 1
+        return block_table
+
+    def add_seqs(
+        self,
+        seqs: List[Sequence],
+        alloc_block_tables: torch.Tensor = None,
+        alloc_block_tables_fn: Callable[[torch.Tensor, torch.Tensor], None] = None,
+    ) -> Union[torch.Tensor, None]:
+        """Add a list of sequences to the batch.
+        User could opt to provide either block tables or a function to allocate block tables.
+
+        Args:
+            seqs (List[Sequence]): The sequences to be added to the batch
+            alloc_block_tables (torch.Tensor): The block tables to be copied and used for the sequence
+            alloc_block_table_fn (Callable[[torch.Tensor, torch.Tensor], None]): The function to allocate blocks for multiple sequences,
+                which is expected to reserve blocks and update status of kv-cache manager.
+
+        Returns:
+            block_tables (torch.Tensor): The block tables of the added sequences, used for block allocation in kv-cache manager.
+                None if the sequences cannot be added.
+        """
+
+        assert (
+            alloc_block_tables is None or alloc_block_tables_fn is None
+        ), "`alloc_block_tables` and `alloc_block_tables_fn` cannot be provided at the same time"
+
+        num_seqs_to_add = min(self.max_batch_size - self._current_batch_size, len(seqs))
+        block_tables = None
+        if num_seqs_to_add > 0:
```

Comment:
How do we deal with the sentences that haven't been added?

Reply:
Note that the input seqs is a list of Sequence. When `num_seqs_to_add` is equivalent with the number of elements in seqs, the function will add all the Sequences and clear the list. When `num_seqs_to_add` is less than the number of elements in seqs, the function will clear the added Sequences in the list, while modify the list to collect Sequences that haven't been added. In other words, in we could check if there exist any Sequences that haven't been added outside the function by checking the seqs list.

I'll revise the docstring later.